### PR TITLE
Removed PH creation from Executor

### LINKED ISSUE #2749: Use Placeholders for intermediate values created by the Partitioner in the Executor
Now ThreadpoolExecutor is creating new Placeholders for intermediate values between `DagNode`s outside of the Module.
https://github.com/pytorch/glow/blob/a8ae908464337eca6216d535c8345e8ed4ed7a8c/lib/Runtime/Executor/ThreadPoolExecutor.cpp#L256-L257
Instead of creating these in the Executor, these should be looked up from the Module.

### PR DESCRIPTION
**Pull request by gcatron**
*Description*: This PR adds a reference to Module in the DAGNode so the executor can use the PHs in the Module rather than needing to create its own.
*Testing*: Updated Executor Test. unit tests pass and resnet-runtime
*Documentation*: Will be added to Runtime doc once runtime doc is added to docs. :D
fixes #2749

### CODE REVIEW

File: tests/unittests/ExecutorTest.cpp [hunk 1]
```diff
@@ -537,28 +544,31 @@ class ExecutorTestBuilder final {
...
-    if (it == placeholders_.end()) {
+    if (!ph) {
       // This is a new symbol. Create a Placeholder and an initialize and new
       // Tensor for it.
-      auto placeholder = llvm::make_unique<Placeholder>(name, type_.get(),
-                                                        /*trainable=*/false);
-      auto *tensor = bindings_->allocate(placeholder.get());
+      auto placeholder = module_->createPlaceholder(type_.get(), name, false);
+      // auto placeholder = llvm::make_unique<Placeholder>(name, type_.get(),
+      //                                                   /*trainable=*/false);
```
Comment (beicy) [inline tests/unittests/ExecutorTest.cpp]: nit: The code can be removed.

File: include/glow/Runtime/RuntimeTypes.h [hunk 2]
```diff
@@ -87,6 +87,10 @@ struct DAGNode {
   /// runtime.
   std::unique_ptr<RuntimeBundle> runtimeBundle;
 
+  /// Pointer to module the function came from. This is so the executor can
+  /// access the associated PHs for the function that are stored in the Module.
+  Module *module;
```
Comment (jackm321) [inline include/glow/Runtime/RuntimeTypes.h]: initialze to nullptr

File: lib/Runtime/Executor/ThreadPoolExecutor.cpp [hunk 3]
```diff
@@ -101,8 +103,10 @@ ExecutionState::ExecutionState(RunIdentifierTy id, const DAGNode *root,
       const auto &symbolInfo = symbolPair.second;
 
       if (symbolInfo.symbolCategory == SymbolCategory::Placeholder) {
-        nodeInputPhBindings->allocate(
-            createOrGetPlaceholder(symbolName, &symbolInfo.type));
+        auto PH = module_->getPlaceholderByName(symbolName);
+        // If a PH name is provided it had to come from the Module originally.
+        DCHECK(PH) << "Placholder: " << symbolName << " is not in the module";
```
Comment (jackm321) [inline lib/Runtime/Executor/ThreadPoolExecutor.cpp]: Nice!

File: lib/Runtime/Executor/ThreadPoolExecutor.h [hunk 4]
```diff
@@ -141,6 +141,10 @@ class ExecutionState final {
   /// Mutex used by bindings insertion functions to make sure only one thread
   /// writes to an ExecutionContext at a time.
   std::mutex bindingsMtx_;
+
+  /// Module for the network. This contains the PHs used by the functions in
+  /// this network.
+  Module *module_;
```
Comment (jackm321) [inline lib/Runtime/Executor/ThreadPoolExecutor.h]: initialize to nullptr

File: lib/Runtime/Executor/ThreadPoolExecutor.cpp [hunk 5]
```diff
@@ -241,27 +245,6 @@ ExecutionContext *ExecutionState::getRawResultContextPtr() const {
   return resultCtx_.get();
 }
 
-Placeholder *ExecutionState::createOrGetPlaceholder(llvm::StringRef name,
```
Comment (SplitInfinity) [inline lib/Runtime/Executor/ThreadPoolExecutor.cpp]: Can you delete the declaration of this from the header too?

File: tests/unittests/ExecutorTest.cpp [hunk 6]
```diff
@@ -222,7 +224,7 @@ class ExecutorTest final {
     // Executor::run() via its callback.
     RunIdentifierTy executorRunId;
     std::unique_ptr<ExecutionContext> executorOutputContext;
-
+    root_->module = module_.get();
```
Comment (SplitInfinity) [inline tests/unittests/ExecutorTest.cpp]: I think this is already done in `ExecutorTestBuilder` on line 308.

File: tests/unittests/ExecutorTest.cpp [hunk 7]
```diff
@@ -537,28 +543,28 @@ class ExecutorTestBuilder final {
   /// mapped for the test being created, reuse the existing value.
   void insertSymbolIntoPlaceholderBindings(llvm::StringRef name,
                                            PlaceholderBindings *bindings) {
-    auto it = placeholders_.find(name);
+    // auto it = placeholders_.find(name);
```
Comment (nadavrot) [inline tests/unittests/ExecutorTest.cpp]: This line should probably be removed or turned into a comment.

File: tests/unittests/ExecutorTest.cpp [hunk 8]
```diff
@@ -201,16 +201,20 @@ class ExecutorTest final {
...
-      : executor_(executor), root_(std::move(root)), type_(std::move(type)),
-        nodes_(std::move(nodes)), placeholders_(std::move(placeholders)),
+      : executor_(executor), root_(std::move(root)), module_(std::move(module)),
+        type_(std::move(type)), nodes_(std::move(nodes)),
+        placeholders_(std::move(placeholders)),
         inputContext_(std::move(inputContext)),
         outputContext_(std::move(outputContext)), runId_(runId),
-        expectSuccess_(expectSuccess), testRun_(false) {}
+        expectSuccess_(expectSuccess), testRun_(false) {
+    root_->module = module_.get();
```
Comment (SplitInfinity) [inline tests/unittests/ExecutorTest.cpp]: I would personally prefer that this be done in `ExecutorTestBuilder`.

File: tests/unittests/ExecutorTest.cpp [hunk 9]
```diff
@@ -222,7 +226,6 @@ class ExecutorTest final {
     // Executor::run() via its callback.
     RunIdentifierTy executorRunId;
     std::unique_ptr<ExecutionContext> executorOutputContext;
-
```
Comment (SplitInfinity) [inline tests/unittests/ExecutorTest.cpp]: Why did you delete this blank line?

Comment (gcatron) [reply tests/unittests/ExecutorTest.cpp]: Unintentionally :-) will undo.

File: tests/unittests/ExecutorTest.cpp (hunk 8 above)
Comment (gcatron) [reply tests/unittests/ExecutorTest.cpp]: I was getting a segfault in the concurrent test when I added it to the builder. Moving it to the test seems to behave properly.

File: lib/Runtime/Executor/ThreadPoolExecutor.cpp (hunk 3 above)
Comment (nickgg) [inline lib/Runtime/Executor/ThreadPoolExecutor.cpp]: nit: Placholder typo

Comment (nickgg) [reply lib/Runtime/Executor/ThreadPoolExecutor.cpp]: I like DCHECK but this is the first time we're using it over GLOW_ASSERT, right?

Comment (gcatron) [reply lib/Runtime/Executor/ThreadPoolExecutor.cpp]: @jackm321 beat me to it, it's already being used in Tensor.cpp as well.
